Document collection subscriptions

hannako opened the pull request:
Navigation and Presentation will soon be rolling out email subscriptions for document collections. 

The required behaviour for these subscriptions is: When a user subscribes to notifications on a document collection, they will be notified when the document collection is updated AND when any of the documents belonging to that collection are updated. 

The relationship between a document collection and the documents it contains is stored as a reverse link. 

This PR adds support for reverse link style lists in the future ([See here for a list of reverse link types](https://github.com/alphagov/publishing-api/blob/main/lib/expansion_rules.rb#L43-L56)), but currently only allows these new lists to be created for document collections.

Related work: https://github.com/alphagov/email-alert-service/pull/583
Trello: https://trello.com/c/tJvdhfdd/1455-support-topic-like-subscriptions-for-document-collections-email-alert-api-m, [Jira issue NAV-5609](https://gov-uk.atlassian.net/browse/NAV-5609)

⚠️  This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

chao-xian reviewed (approved):
LGTM but I'd like for Kevin and Keith to look too.

KludgeKML reviewed (approved):
LGTM

kevindew commented on app/models/subscriber_list_params/single_page_lists.rb:
```diff
@@ -16,7 +18,33 @@ def list_params
+    end
+
+    def reverse_linked_key
+      DOCUMENT_TYPES_PERMITTED_TO_HAVE_REVERSE_LINKED_LISTS[document_type] || ""
```
```suggestion
      DOCUMENT_TYPES_PERMITTED_TO_HAVE_REVERSE_LINKED_LISTS[document_type]
```

I imagine it doesn't matter if this returns nil?

hannako replied:
Ah yes you're right, thank you

kevindew reviewed (commented):
Looks good to me 👍 I've added a few implementation suggestions/problems

hannako commented on app/controllers/single_page_subscriptions_controller.rb:
```diff
@@ -53,15 +53,19 @@ def topic_id
+    content_item_path = params.fetch(:base_path)
+
+    return bad_request unless content_item_path.to_s.starts_with?("/")
+    return bad_request unless URI.parse(content_item_path).relative?
```
This commit isn't necessary for this PR. I was trying to slowly bring the two controllers into alignment, with a view to perhaps refactoring down the line as there is a lot of shared functionality. So this code change was a straight cut and paste. I think for the sake of clarity I will drop this commit, and do any refactoring as a stand alone piece of work. With new specs etc as needed.

hannako commented on app/models/subscriber_list_params/single_page_lists.rb:
```diff
@@ -1,5 +1,7 @@
 module SubscriberListParams
   class SinglePageLists
+    DOCUMENT_TYPES_PERMITTED_TO_HAVE_REVERSE_LINKED_LISTS = { "document_collection" => "document_collections" }.freeze
```
Yes that's better, thank you

hannako committed:
Move the generate_subscriber_list_params_service into a namespaced folder

Changes to the files will be made in the next commit to preserve git history

hannako committed:
Update class names for the namespaced service

hannako committed:
Create a new service to generate list params for single page subscriptions

And add support for reverse linked subscriber lists, only for document
collections

leenagupte commented on app/services/subscriber_list_params/generate_single_page_list_params_service.rb:
```diff
@@ -0,0 +1,47 @@
+        "title" => content_item["title"],
+        "content_id" => content_item["content_id"],
+        "description" => content_item["description"],
+      }.merge(strip_if_empty_value(links))
```
I'm unsure of the need for the `strip_if_empty_value` and `strip_if_empty_key` methods.

I think these checks can probably be done in the `links` method. 

The check here can be replaced with an `present?` check, if the hash is assigned to a variable first.

hannako replied:
Thanks Leena - not sure why I made that so complicated. Have refactored 👍

leenagupte reviewed (commented):
One comment /  question about the "strip" methods.

The rest is looking good. I especially like how the renaming of files and adding of namespaces has been split into separate commits. That made this much easier to review. ⭐

kevindew commented on app/services/subscriber_list_params/generate_links_based_list_params_service.rb:
```diff
@@ -1,51 +1,53 @@
+    attr_reader :content_item
 
-  class UnsupportedContentItemError < StandardError; end
+    class UnsupportedContentItemError < StandardError; end
```
Just semantic, but you should move this out of the private part of this class given it's part of the public interface.

hannako replied:
Thanks, have made that change.

kevindew reviewed (commented):
This looks good to me aside a minor comment. Though I agree with Leena's comment too.

hannako committed:
Make error class public

This is part of this service's public interface.

hannako commented:
@kevindew I've made the changes requested - in the absence of Leena and Kelv, are you happy for me to merge this?
